### ISSUE TITLE
Support for 'timestamp without time zone' type

### ISSUE THREAD

**Issue by omartijn** [yandex/ozo#152]
I am trying to wrap my head around working with timestamps. I have a database that has a field of type 'timestamp without time zone' which I want to query using `ozo`.

The most obvious C++ type for this would be an `std::chrono::system_clock::time_point`. Unfortunately, that didn't immediately work as I had hoped, so I have started looking into how to get this to work.

Some research showed me that a timestamp field in PostgreSQL is basically the number of microseconds since the PostgreSQL epoch - which is defined as the start of this millennia (2000-01-01 00:00:00).

I can get this value out in some sense, I first tried to do `OZO_PG_BIND_TYPE(int64_t, "timestamp");` which fails due to the fact that uint64_t is already taken for something else. I couldn't get this to work using the so-called strong typedef macro, but I got it to bind to `uint64_t` instead - which works as long as the date is in 2000 or later.

If I then use this value in the constructor to `std::chrono::microseconds` and add it to the epoch value I get the right result. What I have trouble is in gluing this together so I can just read the value directly into an `std::chrono::system_clock::time_point`.

Since it is a type from the STL and I can't edit it, I assume I would have to somehow write an adapter function that wraps the conversion going from either postgres to c++ type and vice-versa. I cannot, however, find out how I am expected to do this.

Can anybody give me some pointer as to how to properly integrate this.

**Comment by thed636** [comment]
Hi @omartijn !

Thanks for the issue. There is a limited number of types currently implemented in OZO. The implemented types can be found in [ozo/pg/types](https://github.com/yandex/ozo/tree/master/include/ozo/pg/types) folder (the documentation for the implemented types should be a little bit later). So the `timestamp` is not implemented right now. But yes, you choose the right direction to do not waste your time while waiting for something from us and try to implement it yourself. AFAIR the `timestamp` is transferred as a `double` but not any of integer. So the real serialization and deserialization functions can be written using knowledge of the representation. As a source of such information, I am using [py-proto](https://github.com/MagicStack/py-pgproto). But the PostgreSQL sources is the only native source of such information. So if you want to implement `timestamp` via `std::time_point`, the standard way is:

1. First you need to implement these template specializations:
1.1. [ozo::send_impl\<std::time_point\>](https://yandex.github.io/ozo/structozo_1_1send__impl.html) - to define how to serialize the object,
1.2. [ozo::recv_impl\<std::time_point\>](https://yandex.github.io/ozo/structozo_1_1recv__impl.html) - to define how to deserialize the object,
1.3. [ozo::size_of_impl\<std::time_point\>](https://yandex.github.io/ozo/structozo_1_1size__of__impl.html) - to define how to calculate the size of objects binary representation (which is not really needed as you are using a built-in fixed-size type, so you should not bother with this specialization). 
2. And then make a binding via `OZO_PG_BIND_TYPE(std::time_point, "timestamp")`. 

So this is the sequence. You can do it right in your project source. It is ok, to extend the library via such specializations marked as a customization point. And if everything would be ok you can share your solution via the PR into the library. Please do not hesitate to ask further questions. This part is not covered by how-to and your experience could be a good base for such a chapter.

> I can get this value out in some sense, I first tried to do OZO_PG_BIND_TYPE(int64_t, "timestamp"); which fails due to the fact that uint64_t is already taken for something else. I couldn't get this to work using the so-called strong typedef macro, but I got it to bind to uint64_t instead - which works as long as the date is in 2000 or later.

Could you please to provide an example of code which fails using the `strong typedef`? It should work and this looks like an issue which needs to be fixed.

**Comment by omartijn** [comment]
OK, I have managed to reproduce the issue I had with `OZO_STRONG_TYPEDEF` before. I have used the following two lines to register the type:

```c++
OZO_STRONG_TYPEDEF(int64_t, pg_timestamp);
OZO_PG_BIND_TYPE(pg_timestamp, "timestamp");
```

I then try to read rows into an `ozo::rows_of<int64_t>`, where the row is a selection of the above-mentioned timestamps. This leads to the following run-time error:

> error while processing or converting result from the database, error context: unexpected oid 1114 for type long: no conversion possible from oid to user-supplied type.

If I instead use a `uint64_t` I can get it to work simply by `OZO_PG_BIND_TYPE(uint64_t, "timestamp");`

BTW: About the type being an integer/double: It can actually be either, and it's a compile-time option. Since some time, the default is to send it as an integer. Ideally we should be able to detect this option (maybe it's send during connection setup?) so that we can correctly parse it. I have no idea how to do that, however, and given that it's likely sent as an integer for most people, I'd focus on that now.

**Comment by thed636** [comment]
> I then try to read rows into an `ozo::rows_of<int64_t>`, where the row is a selection of the above-mentioned timestamps.

Well, you should use `ozo::rows_of<pg_timestamp>`, since in case of `ozo::rows_of<int64_t>` the library expected `int8` type oid from Postgres. Oid depends literally on the type you are sending or receiving.

> BTW: About the type being an integer/double: It can actually be either, and it's a compile-time option. Since some time, the default is to send it as an integer. Ideally we should be able to detect this option (maybe it's send during connection setup?) so that we can correctly parse it. I have no idea how to do that, however, and given that it's likely sent as an integer for most people, I'd focus on that now.

I thought about it and decided to do not bother about custom builds with options, because many such options cannot be determined in a normal way. E.g. length of `name` type and so on. So if anybody wants to support such things it should be a compile-time configuration. Because in other cases it would affect a runtime and it is not good - common users should not pay for rare customizations.

**Comment by omartijn** [comment]
This issue was solved by #155
